File: EvoPGA/codes/PGA.go
// PGA.go
/* To annotate PGAG or RAST gene annotation table with Salmonella PG UID for each protein gene. Here, we use 26-genome Salmonella PG to make the annotation.
   Procedure: (1) Making the alignment between each proteome of the 26 Salmonella strains and the proteome to be annotated;
              (2) Finding the mutual best alignments for each strain and the one to be annotated;
              (3) Using "26_PG.txt" to annotate the genes of targeted strains with PG UIDs; Some genes could not be annotated because of no PG cluster was
                  mapped or pseudogenes / non-coding RNAs.
   USAGE:     Copy the compiled program, "26_PG.txt" and "S.genus.AG_PGAG.gene.tab.txt" or "Salmonella.genus.ancient.orthologous.genome_RAST.tab.txt" into
              ~/BactCG1.0/results/out_mutbest_filt/, running the command below.

   Command:   ./PGA  <26_PG.txt>    <PGAG|RAST_tab.txt>   <PGAG|RAST>   <Strain_to_be_annotated>
*/

package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

func main() {
	if len(os.Args) < 5 {
		fmt.Printf("Errors in inputing files!\n")
		return
	}
	pgmapmatFileName := os.Args[1]
	pgmapFile, openErr1 := os.Open(pgmapmatFileName)
	if openErr1 != nil {
		fmt.Printf("Error happened in opening %s!\n", pgmapmatFileName)
		return
	}
	defer pgmapFile.Close()

	pgmapReader := bufio.NewReader(pgmapFile)
	pg := make(map[string]string)
	for {
		pgmap, readErr1 := pgmapReader.ReadString('\n')
		pgmap = seqTrim(pgmap)
		if len(pgmap) > 0 {
			pgmapArray := strings.Split(pgmap, "\t")
			//fmt.Printf("%d\n",len(pgmapArray))
			for i := 2; i < len(pgmapArray)-1; i++ {
				if !(pgmapArray[i] == "-") {
					pg[pgmapArray[i]] = pgmapArray[0]
					//fmt.Printf("%s\t%s\n",pgmapArray[i],pg[pgmapArray[i]])
				}
			}
		}
		if readErr1 == io.EOF {
			break
		}
	}

	var flag int
	if strings.Contains(os.Args[3], "PGAG") {
		flag = 2
	} else if strings.Contains(os.Args[3], "RAST") {
		flag = 1
	}

	strain := os.Args[4]

	annMap := make(map[string]string)

	dir := "/home/yaozikun/Public_Dir/Bact_PGA/BactPGA/test/seq"
	entries := os.ReadDir(dir)

	count := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		if filepath.Ext(name) != ".faa" {
			continue
		}
		base := name[:len(name)-4]
		if _, err := strconv.Atoi(base); err == nil {
			count++
		}
	}

	for i := 1; i <= count; i++ {
		mutbestFileName := strain + ".vs." + strconv.Itoa(i) + ".mutbest.filt.txt"
		mutbestFile, openErr2 := os.Open(mutbestFileName)
		if openErr2 != nil {
			fmt.Printf("Error happened in opening %s!\n", mutbestFileName)
			return
		}
		defer mutbestFile.Close()

		mutbestReader := bufio.NewReader(mutbestFile)

		for {
			mutbest, readErr2 := mutbestReader.ReadString('\n')
			mutbest = seqTrim(mutbest)
			if len(mutbest) > 0 {
				mutbestArray := strings.Split(mutbest, "\t")
				if !(len(annMap[mutbestArray[0]]) > 0) {
					if len(pg[mutbestArray[1]]) > 0 {
						annMap[mutbestArray[0]] = pg[mutbestArray[1]]
						//fmt.Printf("%s\t%s\n",mutbestArray[0],annMap[mutbestArray[0]])
					}
				}

			}
			if readErr2 == io.EOF {
				break
			}
		}
	}

	agFileName := os.Args[2]
	agFile, openErr3 := os.Open(agFileName)
	if openErr3 != nil {
		fmt.Printf("Error happened in opening %s!\n", agFileName)
		return
	}
	defer agFile.Close()

	agReader := bufio.NewReader(agFile)
	for {
		ag, readErr3 := agReader.ReadString('\n')
		ag = seqTrim(ag)
		if len(ag) > 0 {
			agArray := strings.Split(ag, "\t")
			for i := 0; i <= flag; i++ {
				fmt.Printf("%s\t", agArray[i])
			}
			if len(annMap[agArray[flag]]) > 0 {
				fmt.Printf("%s", annMap[agArray[flag]])
			}
			for i := flag + 1; i < len(agArray); i++ {
				fmt.Printf("\t%s", agArray[i])
			}
			fmt.Printf("\n")
		}
		if readErr3 == io.EOF {
			break
		}
	}
}

func seqTrim(s string) string {
	s = strings.Trim(s, "\r\n")
	s = strings.Trim(s, "\n")
	//   s = strings.Replace(s," ","",-1)
	return s
}

File: EvoPGA/codes/PGA.py
#!/usr/bin/env python3
"""
PGA.py
======

Python reimplementation of PGA.go

Purpose:
    Annotate PGAG or RAST gene annotation table with Salmonella PG UID for each protein gene.
    Uses 26-genome Salmonella PG for annotation.

Procedure:
    1. Align each proteome of 26 Salmonella strains with the target strain.
    2. Find mutual best alignments.
    3. Use "26_PG.txt" to annotate genes with PG UIDs.

Usage:
    python PGA.py <26_PG.txt> <PGAG_or_RAST_tab.txt> PGAG|RAST <StrainName> [rootDir]

Example:
    python PGA.py 26_PG.txt S.genus.AG_PGAG.gene.tab.txt PGAG StrainName
"""

import os
import sys 

def seq_trim(s: str) -> str:
    return s.strip('\r\n')

def load_pg_map(pg_file: str) -> dict:
    pg = {}
    with open(pg_file, 'r') as f:
        for line in f:
            line = seq_trim(line)
            if not line:
                continue
            parts = line.split('\t')
            pg_uid = parts[0]
            for i in range(2, len(parts) - 1):
                gene = parts[i]
                if gene != '-':
                    pg[gene] = pg_uid
    return pg

def load_mutbest_files(strain: str, count: int, root: str) -> dict:
    ann_map = {}
    for i in range(1, count + 1):
        filename = f"{strain}.vs.{i}.best.filt.txt"
        filepath = os.path.join(root, filename)
        if not os.path.isfile(filepath):
            print(f"Error: Cannot open {filepath}")
            sys.exit(1)

        with open(filepath, 'r') as f:
            for line in f:
                line = seq_trim(line)
                if not line:
                    continue
                parts = line.split('\t')
                query = parts[0]
                subject = parts[1]
                if query not in ann_map:
                    ann_map[query] = subject
    return ann_map

def count_faa_files(seq_dir: str) -> int:
    count = 0
    for fname in os.listdir(seq_dir):
        if fname.endswith('.faa'):
            try:
                int(fname[:-4])
                count += 1
            except ValueError:
                continue
    return count

def annotate_and_print(ag_file: str, ann_map: dict, pg: dict, flag: int):
    """Read annotation file and print annotated output."""
    with open(ag_file, 'r') as f:
        for line in f:
            line = seq_trim(line)
            if not line:
                continue
            parts = line.split('\t')
            gene = parts[flag]
            pg_uid = pg.get(ann_map.get(gene), '')

            print('\t'.join(parts[:flag + 1]), end='')
            print(f"\t{pg_uid}", end='')
            if len(parts) > flag + 1:
                print(f"\t" + '\t'.join(parts[flag + 1:]))
            else:
                print()

def main():
    if len(sys.argv) < 5:
        print("Usage: python PGA.py <255_PG.txt> <PGAG_or_RAST_tab.txt> PGAG|RAST <StrainName> [rootDir]")
        sys.exit(1)

    root = "/home/yaozikun/Public_Dir/Bact_PGA/BactPGA/input"
    if len(sys.argv) >= 6:
        root = sys.argv[5]

    pg_file = os.path.join(root, sys.argv[1])
    ag_file = os.path.join(root, sys.argv[2])
    strain = sys.argv[4]
    mode = sys.argv[3].upper()

    if mode == "PGAG":
        flag = 2
    elif mode == "RAST":
        flag = 1
    else:
        print("Third argument must be PGAG or RAST")
        sys.exit(1)

    seq_dir = os.path.join(root, "seq")
    if not os.path.isdir(seq_dir):
        print(f"Error: seq directory not found at {seq_dir}")
        sys.exit(1)

    count = count_faa_files(seq_dir)

    pg = load_pg_map(pg_file)
    ann_map = load_mutbest_files(strain, count, root)
    annotate_and_print(ag_file, ann_map, pg, flag)

if __name__ == "__main__":
    main()